streams/shared/src/main/scala/zio/stream/ZPipeline.scala

```diff
@@ -1820,22 +1820,23 @@ object ZPipeline extends ZPipelinePlatformSpecificConstructors {
         chunk => {
           val size = chunk.size
 
-          if (size == 0) reader
-          else if (size == 1) {
+          if (size == 1) {
             val a = chunk.head
 
             f(a) match {
               case r: Right[?, Out] => ZChannel.write(Chunk.single(r.value)) *> reader
               case l: Left[Err, ?]  => ZChannel.refailCause(Cause.fail(l.value))
             }
           } else {
-            val builder: ChunkBuilder[Out] = ChunkBuilder.make[Out](chunk.size)
-            val iterator: Iterator[In]     = chunk.iterator
+            val builder: ChunkBuilder[Out] = ChunkBuilder.make[Out](size)
+            val iterator                   = chunk.chunkIterator
+            var index: Int                 = 0
             var error: Err                 = null.asInstanceOf[Err]
 
-            while (iterator.hasNext && (error == null)) {
-              val a = iterator.next()
-              f(a) match {
+            while (iterator.hasNextAt(index) && (error == null)) {
```

Comment:
I'm 99.99% sure that if you have `size` already stored locally you can simply iterate by checking `index < size` without having to call `hasNextAt`

Reply:
Done! Sorry for the delay